### DISCUSSION TITLE
Pandino - An OSGi - lite framework for JavaScript runtimes.

### DISCUSSION THREAD

**Discussion by as-shura** [eclipse-theia/theia#12669] [Improvements] [answered]
I was wondering if it is possible to integrate Pandino as the base framework inside of Theia? Since Eclipse is built around OSGi why not stay in this paradigm even in the base framework foundation.

https://github.com/BlackBeltTechnology/pandino

**Comment by paul-marechal** [comment]
It feels to me that this library fills the same role as Inversify? What would be key functionality differences?

We use Inversify so extensively that I don't think we can change inversion of control framework without an extremely compelling reason...

"Eclipse (framework) did it so why not Theia?" is not a valid argument IMO and the answer is that Pandino wasn't around when Theia was initially created back in [Feb 2017](https://github.com/eclipse-theia/theia/commit/c20fed2b5045392e113071176a284addfbde84fa).

**Reply by as-shura** [reply]
I thought that learning Inversify is adding one extra step in the learning curve and goes a bit unconventional to the people who already know the Eclipse Plugin framework.

To keep the feeling the same I though this framework offers the same development feeling.

Well my argument might not be that reasonable it was just an idea.

**Reply by paul-marechal** [reply]
I only know how to use Inversify and not OSGi ;)

Marking as solved unless you want to bring more arguments in favor.

**Reply by as-shura** [reply]
Well I started a thread with the intent to combine all the features from OSGi inside of Javascript. The authors of many OSGi like DI container libraries put the effort to come up with same features but I think that taking other's earlier work can be used and implemented if many functionalities are missing.

https://github.com/BlackBeltTechnology/pandino/issues/102

If you don't know OSGi and want to see what kind of features it has then you can look at what is available in the state of the art modular UI framework like

https://github.com/Drombler/drombler-fx

You can look at how they are building the UI with annotations and modularity

https://www.drombler.org/drombler-fx/1.0/docs/tutorial/index.html#tutorial

**Reply by paul-marechal** [reply]
> if many functionalities are missing

Such as?

**Reply by msujew** [reply]
Yes, the Theia plugin system (used for VS Code extensions) is designed to allow to add code during runtime. The Theia extension system is compile-time, the plugin system is runtime.

**Reply by msujew** [reply]
Uhm, Inversify can also change its bindings on the fly - you can dynamically rebind or unbind symbols as required. Theia extensions are by design compile-time, not due to a technical limitation. Both frameworks likely cover the same feature set. Using a different DI library is a huge undertaking that likely none of the established contributors will support.

**Reply by fayakun-it-consulting** [reply]
> Uhm, Inversify can also change its bindings on the fly - you can dynamically rebind or unbind symbols as required. Theia extensions are by design compile-time, not due to a technical limitation. Both frameworks likely cover the same feature set. Using a different DI library is a huge undertaking that likely none of the established contributors will support.

That's not true because you as the programmer have to input that code where you would like to make it dynamic. There's no automatic IoC management such as Pandino SCR.

**Comment by paul-marechal** [comment, marked as answer]
@eclipse-theia/core correct me if I'm wrong:

It would seem like there is currently no interest from all the major Eclipse Theia contributors to refactor the whole code base to use Pandino. The only traction comes from @as-salam and @fayakun-it-consulting which haven't contributed to Theia's sources/features/bug-fixes as of yet.

Couple of reasons why the Theia project may not adopt Pandino at this time:

- Unknown and seemingly unused library at large: low download counts and low amount of public dependents (actually 0).
- Massive endeavour: Integrating this library would affect the whole code base.
- Massive API breakage: This will most likely require all Theia extenders to update their code bases as well.
- Missing broader interest: outside of the current requesters, not enough contributing parties expressed interest in this.
